File: task_templates/3_pipelines/4_python3_xgboost/create_pipeline.py
"""
Copyright 2021 DataRobot, Inc. and its affiliates.
All rights reserved.
This is proprietary source code of DataRobot, Inc. and its affiliates.
Released under the terms of DataRobot Tool and Utility Agreement.
"""
"""
    This is a helper file for the custom.py main file
    which actually creates the pipeline, which consists of data preprocessing steps and an XGBoost model

    Note: This script can build either a classification or regression model depending on the target

"""
from xgboost import XGBClassifier, XGBRegressor

from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

import pandas as pd


def create_regression_model() -> XGBRegressor:
    """
    Create a regression model.

    Returns
    -------
    XGBRegressor
        XGBoost regressor model
    """
    xg_reg = XGBRegressor(
        objective="reg:squarederror",
        learning_rate=0.05,
        max_depth=5,
        min_child_weight=20,
        n_estimators=50,
        seed=123,
    )
    return xg_reg


def create_binary_classification_model() -> XGBClassifier:
    """
    Create a binary classification model.

    Returns
    -------
    XGBClassifier
        XGBoost classifier model
    """
    xg_clf = XGBClassifier(
        objective="binary:logistic",
        learning_rate=0.05,
        max_depth=5,
        min_child_weight=20,
        n_estimators=50,
        seed=123,
    )
    return xg_clf


def create_multiclass_classification_model(num_labels: int) -> XGBClassifier:
    """
    Create a binary classification model.

    Parameters
    ----------
    num_labels
        number of classification labels

    Returns
    -------
    XGBClassifier
        XGBoost classifier model
    """
    xg_clf = XGBClassifier(
        objective="multi:softprob",
        num_class=num_labels,
        learning_rate=0.05,
        max_depth=5,
        min_child_weight=20,
        n_estimators=50,
        seed=123,
    )
    return xg_clf


def make_classifier_pipeline(X: pd.DataFrame, num_labels: int) -> Pipeline:
    """
    Make the classifier pipeline with the required preprocessor steps and estimator in the end.

    Parameters
    ----------
    X
        X containing all the required features for training
    num_labels
        number of classification labels

    Returns
    -------
    Pipeline
        Classifier pipeline with preprocessor and estimator
    """
    numerics = ["int16", "int32", "int64", "float16", "float32", "float64"]
    num_features = list(X.dropna(axis=1, how="all").select_dtypes(include=numerics).columns)

    # This example model only uses numeric features and drops the rest
    num_transformer = Pipeline(
        steps=[("imputer", SimpleImputer(strategy="median")), ("scaler", StandardScaler())]
    )

    preprocessor = ColumnTransformer(transformers=[("num", num_transformer, num_features)])

    # create model
    estimator = (
        create_binary_classification_model()
        if num_labels == 2
        else create_multiclass_classification_model(num_labels)
    )

    # pipeline with preprocessor and estimator bundled
    classifier_pipeline = Pipeline(steps=[("preprocessor", preprocessor), ("estimator", estimator)])
    return classifier_pipeline


def make_regressor_pipeline(X: pd.DataFrame) -> Pipeline:
    """
    Make the regressor pipeline with the required preprocessor steps and estimator in the end.

    Parameters
    ----------
    X
        X containing all the required features for training

    Returns
    -------
    Pipeline
        Regressor pipeline with preprocessor and estimator
    """
    numerics = ["int16", "int32", "int64", "float16", "float32", "float64"]
    num_features = list(X.dropna(axis=1, how="all").select_dtypes(include=numerics).columns)

    # This example model only uses numeric features and drops the rest
    num_transformer = Pipeline(
        steps=[("imputer", SimpleImputer(strategy="mean")), ("standardize", StandardScaler())]
    )
    preprocessor = ColumnTransformer(transformers=[("num", num_transformer, num_features)])

    # create model
    estimator = create_regression_model()

    # pipeline with preprocessor and estimator bundled
    regressor_pipeline = Pipeline(steps=[("preprocessor", preprocessor), ("estimator", estimator)])
    return regressor_pipeline

File: task_templates/3_pipelines/3_r_lang/create_pipeline.R
# Copyright 2021 DataRobot, Inc. and its affiliates.
#
# All rights reserved.
#
# 
#
# This is proprietary source code of DataRobot, Inc. and its affiliates.
#
# Released under the terms of DataRobot Tool and Utility Agreement.
create_pipeline<-function(X, y, model_type='regression') {
  # set up dataframe for modeling
  train_df <- X
  train_df$target <- unlist(y)
  if (model_type == 'classification'){
    train_df$target <- as.vector(as.factor(train_df$target))
  }

  # set up the modeling pipeline
  model_recipe <- recipe(target ~ ., data = train_df) %>%
    # Drop constant columns
    step_zv(all_predictors()) %>%
    # Numeric preprocessing
    step_normalize(all_numeric(), -all_outcomes()) %>%
    # Categorical preprocessing
    step_other(all_nominal(), -all_outcomes()) %>%
    step_dummy(all_nominal(), -all_outcomes())

    gbmGrid <- expand.grid(
      n.trees = 50,
      interaction.depth=1,
      shrinkage = 0.1,
      n.minobsinnode = 10
    )

    # Run the model using caret
    set.seed(1234)
    model <- train(
      model_recipe,
      train_df,
      method = "gbm",
      trControl = trainControl(method = "cv", number = 3, seeds=list(c(2,2,2), c(2,2,2), c(2,2,2), c(2))),
      tuneGrid = gbmGrid
    )

    return(model)
}
